### PR TITLE
feat: add MergePathAndValueIntoMap

### PR DESCRIPTION
**Pull request by naveensrinivasan**
## Description

- included a helper for generating a nested map.

## Related Issue
https://github.com/defenseunicorns/zarf/pull/2403#discussion_r1543576325

### CODE REVIEW

File: helpers/misc.go [hunk 1]
```diff
@@ -146,3 +147,28 @@ func MergeNonZero[T any](original T, overrides T) T {
...
+			current[part] = value
+		} else {
+			if _, exists := current[part]; !exists {
+				// If the part does not exist, create a new map for it.
+				current[part] = make(map[string]any)
+			}
+
+			nextMap, ok := current[part].(map[string]any)
+			if !ok {
+				return fmt.Errorf("conflict at '%s', expected map but got %T", strings.Join(pathParts[:i+1], "."), current[part])
```
Comment (Noxsios) [inline helpers/misc.go]:
```suggestion
				return fmt.Errorf("conflict at %q, expected map[string]any but got %T", strings.Join(pathParts[:i+1], "."), current[part])
```

`%q` will wrap strings in `"` for you, anything that implements Stringers and can be a `%s` can also be a `%q`.

Usage varies depending upon which one you want to use in a string.

File: helpers/misc_test.go [hunk 2]
```diff
@@ -199,3 +200,89 @@ func (suite *TestMiscSuite) TestBoolPtr() {
...
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			err := MergePathAndValueIntoMap(tt.args.m, tt.args.path, tt.args.value)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("MergePathAndValueIntoMap() error = %v, wantErr %v", err, tt.wantErr)
+			}
+
+			if !reflect.DeepEqual(tt.args.m, tt.want) {
+				t.Errorf("Map structure mismatch: got %v, want %v", tt.args.m, tt.want)
+			}
```
Comment (Noxsios) [inline helpers/misc_test.go]: use `require` here, see other `_test.go` files for examples

File: helpers/misc.go [hunk 3]
```diff
@@ -146,3 +147,28 @@ func MergeNonZero[T any](original T, overrides T) T {
 	}
 	return originalValue.Elem().Interface().(T)
 }
+
+// MergePathAndValueIntoMap takes a path in dot notation as a string and a value (also as a string for simplicity),
+// then merges this into the provided map. The value can be any type.
+func MergePathAndValueIntoMap(m map[string]any, path string, value any) error {
```
Comment (lucasrod16) [inline helpers/misc.go:153]: The function signature may need to be changed depending on whether Zarf chart variables will have a `value:` field that is included in the merge behavior. If a `value:` field is needed, then that may need to be added as an additional argument to this function.

https://github.com/defenseunicorns/zarf/pull/2403/files#r1543588026
https://github.com/defenseunicorns/zarf/pull/2403/files#r1543612278

File: helpers/misc_test.go [hunk 4]
```diff
@@ -199,3 +200,89 @@ func (suite *TestMiscSuite) TestBoolPtr() {
 func TestMisc(t *testing.T) {
 	suite.Run(t, new(TestMiscSuite))
 }
+
+func TestMergePathAndValueIntoMap(t *testing.T) {
```
Comment (Noxsios) [inline helpers/misc_test.go]: add this test to the currently existing test suite in this file

**Review by Noxsios — CHANGES_REQUESTED** [review]

File: helpers/misc.go (hunk 3 above)
Comment (lucasrod16) [reply helpers/misc.go:153]: a `value:` field is not needed: https://github.com/defenseunicorns/zarf/pull/2403#discussion_r1543715691

File: helpers/misc.go (hunk 1 above)
Comment (naveensrinivasan) [reply helpers/misc.go]: I took care of it.

File: helpers/misc_test.go (hunk 2 above)
Comment (naveensrinivasan) [reply helpers/misc_test.go]: I took care of it. Thanks

File: helpers/misc_test.go (hunk 4 above)
Comment (naveensrinivasan) [reply helpers/misc_test.go]: I did that. Thanks

File: helpers/misc.go [hunk 5]
```diff
@@ -146,3 +147,28 @@ func MergeNonZero[T any](original T, overrides T) T {
...
+			current[part] = value
+		} else {
+			if _, exists := current[part]; !exists {
+				// If the part does not exist, create a new map for it.
+				current[part] = make(map[string]any)
+			}
+
+			nextMap, ok := current[part].(map[string]any)
+			if !ok {
+				return fmt.Errorf("conflict at '%q', expected map but got %T", strings.Join(pathParts[:i+1], "."), current[part])
```
Comment (lucasrod16) [inline helpers/misc.go]:
```suggestion
				return fmt.Errorf("conflict at %q, expected map but got %T", strings.Join(pathParts[:i+1], "."), current[part])
```

**Review by lucasrod16 — CHANGES_REQUESTED** [review]

**Review by lucasrod16 — APPROVED** [review]

**Review by Noxsios — APPROVED** [review]

**Pull request merged**